test/unittest/tests/test_qsort.cpp

```diff
@@ -0,0 +1,173 @@
+    random_generator &operator=(const random_generator &) = delete;
+    ~random_generator() = default;
+
+    std::vector<int>
+    make_vector(size_t size)
+    {
+        std::vector<int> rv(size);
+        auto rnd = [this]() { return this->dist(this->mersenne_engine); };
+        std::generate(begin(rv), end(rv), rnd);
+        return rv;
+    }
+
+    std::vector<compound_test_type>
+    make_compound_vector(size_t size)
+    {
+        std::vector<compound_test_type> rv(size);
+        auto rnd = [this]() { return compound_test_type(this->dist(this->mersenne_engine)); };
+        std::generate(begin(rv), end(rv), rnd);
+
+        return rv;
+    }
+};
+
+int
+simple_cmp(const void *a, const void *b)
+{
+    auto lhs = static_cast<const int *>(a);
+    auto rhs = static_cast<const int *>(b);
+
+    return *lhs - *rhs;
+}
+
+int
+compound_cmp(const void *a, const void *b)
+{
+    auto lhs = *(static_cast<const compound_test_type *>(a));
+    auto rhs = *(static_cast<const compound_test_type *>(b));
+
+    return lhs.a - rhs.a;
+}
+
+int
+transposable_cmp(const void *a, const void *b, void *ctx)
+{
+    bool reverse = ctx ? *(static_cast<bool *>(ctx)) : false;
+    auto lhs = static_cast<const int *>(a);
+    auto rhs = static_cast<const int *>(b);
+
+    if (reverse)
+        return *rhs - *lhs;
+    return *lhs - *rhs;
+};
+
+int
+counting_cmp(const void *a, const void *b, void *ctx)
+{
+    auto count = static_cast<int *>(ctx);
+    (*count)++;
+
+    auto lhs = static_cast<const int *>(a);
+    auto rhs = static_cast<const int *>(b);
+    return *lhs - *rhs;
+};
+} // namespace
+
+TEST_CASE("Safe to invoke on an empty array", "[qsort]")
+{
+    std::vector<int> input;
+    __wt_qsort_r(&input[0], input.size(), sizeof(input[0]), transposable_cmp, nullptr);
+    __wt_qsort(&input[0], input.size(), sizeof(input[0]), simple_cmp);
+}
+
+TEST_CASE("Single element", "[qsort]")
+{
+    std::vector<int> input = {123};
+    __wt_qsort_r(&input[0], input.size(), sizeof(input[0]), transposable_cmp, nullptr);
+    CHECK(input[0] == 123);
+}
+
+TEST_CASE("Compound type", "[qsort]")
+{
+    random_generator rand_gen;
+    std::vector<compound_test_type> input{rand_gen.make_compound_vector(1000)};
+
+    __wt_qsort(&input[0], input.size(), sizeof(input[0]), compound_cmp);
+
+    CHECK(std::is_sorted(input.begin(), input.end()));
+    for (int i = 0; i < 1000; i++) {
+        CHECK(input[i].b == compound_test_type::COMPOUND_MAGIC_B);
+        CHECK(input[i].c == compound_test_type::COMPOUND_MAGIC_C);
+    }
+}
+
+TEST_CASE("Check contents", "[qsort]")
+{
+    random_generator rand_gen;
+    std::vector<int> input{rand_gen.make_vector(10000)};
+    std::vector<int> input_copy{input};
+
+    __wt_qsort(&input[0], input.size(), sizeof(input[0]), simple_cmp);
+
+    CHECK(std::is_sorted(input.begin(), input.end()));
+    for (int i = 0; i < 10000; i++)
+        CHECK(std::find(input_copy.begin(), input_copy.end(), input[i]) != input_copy.end());
```

Comment:
Realized `std::set` wouldn't work here as it doesn't tolerate duplicates. Nor will this handle duplicates correctly.
Suggest something like:
```
std::list input_copy(input.begin(), input.end());
// for each element in input
  auto match = std::find(input_copy.begin(), input_copy.end(), input[i]);
  CHECK(match != input_copy.end());
  input_copy.erase(match);
```

Reply:
Good point about dupes - fixed in 02f0c61.